pennylane/pauli/pauli_arithmetic.py

```diff
@@ -0,0 +1,261 @@
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""The Pauli arithmetic abstract reduced representation classes"""
+from copy import copy
+from functools import reduce
+
+import numpy as np
+from scipy import sparse
+from pennylane import math, wires
+
+I = "I"
+X = "X"
+Y = "Y"
+Z = "Z"
+
+matI = np.eye(2)
+matX = np.array([[0, 1], [1, 0]])
+matY = np.array([[0, -1j], [1j, 0]])
+matZ = np.array([[1, 0], [0, -1]])
+
+mat_map = {
+    I: matI,
+    X: matX,
+    Y: matY,
+    Z: matZ,
+}
+
+sparse_matI = sparse.eye(2, format="csr")
+sparse_matX = sparse.csr_matrix([[0, 1], [1, 0]])
+sparse_matY = sparse.csr_matrix([[0, -1j], [1j, 0]])
+sparse_matZ = sparse.csr_matrix([[1, 0], [0, -1]])
+
+sparse_mat_map = {
+    I: sparse_matI,
+    X: sparse_matX,
+    Y: sparse_matY,
+    Z: sparse_matZ,
+}
+
+_map_I = {
+    I: (1, I),
+    X: (1, X),
+    Y: (1, Y),
+    Z: (1, Z),
+}
+_map_X = {
+    I: (1, X),
+    X: (1, I),
+    Y: (1.0j, Z),
+    Z: (-1.0j, Y),
+}
+_map_Y = {
+    I: (1, Y),
+    X: (-1.0j, Z),
+    Y: (1, I),
+    Z: (1j, X),
+}
+_map_Z = {
+    I: (1, Z),
+    X: (1j, Y),
+    Y: (-1.0j, X),
+    Z: (1, I),
+}
+
+mul_map = {I: _map_I, X: _map_X, Y: _map_Y, Z: _map_Z}
+
+
+class PauliWord(dict):
+    """Immutable dictionary used to represent a Pauli Word.
+    Can be constructed from a standard dictionary.
+
+    >>> w = PauliWord({"a": X, 2: Y, 3: Z})
+    """
+
+    def __missing__(self, key):
+        """If the wire is not in the Pauli word,
+        then no operator acts on it, so return the Identity."""
+        return I
+
+    def __init__(self, mapping):
+        """Strip identities from PauliWord on init!"""
+        mapping = {wire: op for wire, op in mapping.items() if op != I}
+        super().__init__(mapping)
+
+    def __copy__(self):
+        """Copy the PauliWord instance."""
+        return PauliWord(dict(self.items()))
+
+    def __setitem__(self, key, item):
+        """Restrict setting items after instantiation."""
+        raise NotImplementedError
+
+    def update(self, __m, **kwargs) -> None:
+        """Restrict updating PW after instantiation."""
+        raise NotImplementedError
+
+    def __hash__(self):
+        return hash(frozenset(self.items()))
+
+    def __mul__(self, other):
+        """Multiply two Pauli words together using the matrix product if wires overlap
+        and the tensor product otherwise.
+
+        Args:
+            other (PauliWord): The Pauli word to multiply with
+
+        Returns:
+            result(PauliWord): The resulting operator of the multiplication
+            coeff(complex): The complex phase factor
+        """
+        c_self, c_other = (copy(self), copy(other))
```

Comment:
Why do we copy both PWs? We are copying a lot of Pauli ops that will be modified/removed afterwards.

I believe it would be faster if we created an empty dict, populated it with the multiplication results and then casted it to a `PauliWord` instead of doing these 2 copies

Reply:
The copies were added after the fact. Initially, the idea was that if we allow for one of the inputs to get mutated, then we can compute the multiplication in $~ O(min(m, n))$ time, where m, n are the number of factors in each Pauliword respectively. By adding the copy, the complexity becomes: $~ O(max(m, n))$. I was able to remove one of the two copies, but the time complexity remains the same. If we were to populate an empty dict, we would have to iterate over both PWs and keep track of terms that have already been multiplied. This would end up being the same time complexity $~ O(max(m, n))$. For this reason I think we should stick with this implementation.